### ISSUE TITLE
transformer exception: console buffer yet initialized

### ISSUE THREAD

**Issue by ewot** [AdalbertoLacruz/less_dart#8]
When running as a transformer from inside the Eclipse plugin on Ubuntu I frequently get the following error:

```
Build error:
Transform File on xxx|less threw error: Bad state: Console buffer yet initialized
package:less_dart/src/logger.dart 37                       Logger.Logger
package:less_dart/less.dart 37                             Less.Less
package:less_dart/srcTransformer/less_transformer.dart 30  LessTransformer.transform.<fn>
dart:async/zone.dart 895                                   _rootRun
dart:async/zone.dart 796                                   _CustomZone.run
dart:async/zone.dart 1251                                  runZoned
package:less_dart/srcTransformer/less_transformer.dart 29  LessTransformer.transform
package:less_dart/transformer.dart 90                      FileTransformer.apply.<fn>.<fn>.<fn>
dart:async/zone.dart 1155                                  _RootZone.runUnary
dart:async/future_impl.dart 494                            _Future._propagateToListeners.handleValueCallback
dart:async/future_impl.dart 577                            _Future._propagateToListeners
dart:async/future_impl.dart 368                            _Future._completeWithValue
dart:async/future_impl.dart 422                            _Future._asyncComplete.<fn>
dart:async/schedule_microtask.dart 41                      _asyncRunCallbackLoop
dart:async/schedule_microtask.dart 48                      _asyncRunCallback
dart:isolate-patch/isolate_patch.dart 96                   _runPendingImmediateCallback
dart:isolate-patch/isolate_patch.dart 143                  _RawReceivePortImpl._handleMessage

package:less_dart/src/logger.dart 37                       Logger.Logger
package:less_dart/less.dart 37                             Less.Less
package:less_dart/srcTransformer/less_transformer.dart 30  LessTransformer.transform.<fn>
dart:async                                                 runZoned
package:less_dart/srcTransformer/less_transformer.dart 29  LessTransformer.transform
package:less_dart/transformer.dart 90                      FileTransformer.apply.<fn>.<fn>.<fn>
dart:isolate                                               _RawReceivePortImpl._handleMessage
```

This is how it is configured in pubspec.yaml:

```
transformers:
- less_dart:
    silence: true
    no-color: true
- angular
- $dart2js:
    sourceMaps: true
```

**Comment by AdalbertoLacruz** [comment]
I need more information.
Remove silence flag, to know was happening inside.

By the way, "no-color" flag is not supported in the transformer.

**Comment by ewot** [comment]
I removed the silence flag but didn't get much more info. Full trace last in the the comment.

In this case some files are compiled successfully but then it crashes with a few files left.

Looks like some kind of race condition when it initilizes the logger.

This is the output:

```
[Info from less-dart] lessc --no-color web/less/variables.less > web/less/variables.css took 0.0s
[Info from less-dart] lessc --no-color web/less/mixins.less > web/less/mixins.css took 0.0s
[Info from less-dart] lessc --no-color web/less/paginator.less > web/less/paginator.css took 0.0s
[Info from less-dart] lessc --no-color web/less/wallet.less > web/less/wallet.css took 0.0s
[Info from less-dart] lessc --no-color web/less/modal.less > web/less/modal.css took 0.0s
[Info from less-dart] lessc --no-color web/less/customization.less > web/less/customization.css took 0.0s
Build error:
Transform File on xxx|less threw error: Bad state: Console buffer yet initialized
package:less_dart/src/logger.dart 37:7                        Logger.Logger
package:less_dart/less.dart 37:18                             Less.Less
package:less_dart/srcTransformer/less_transformer.dart 30:23  LessTransformer.transform.<fn>
dart:async/zone.dart 895                                      _rootRun
dart:async/zone.dart 796                                      _CustomZone.run
dart:async/zone.dart 1251                                     runZoned
package:less_dart/srcTransformer/less_transformer.dart 29:5   LessTransformer.transform
package:less_dart/transformer.dart 90:32                      FileTransformer.apply.<fn>.<fn>.<fn>
dart:async/zone.dart 1155                                     _RootZone.runUnary
dart:async/future_impl.dart 494                               _Future._propagateToListeners.handleValueCallback
dart:async/future_impl.dart 577                               _Future._propagateToListeners
dart:async/future_impl.dart 368                               _Future._completeWithValue
dart:async/future_impl.dart 422                               _Future._asyncComplete.<fn>
dart:async/schedule_microtask.dart 41                         _asyncRunCallbackLoop
dart:async/schedule_microtask.dart 48                         _asyncRunCallback
dart:isolate-patch/isolate_patch.dart 96                      _runPendingImmediateCallback
dart:isolate-patch/isolate_patch.dart 143                     _RawReceivePortImpl._handleMessage

package:less_dart/src/logger.dart 37:7                        Logger.Logger
package:less_dart/less.dart 37:18                             Less.Less
package:less_dart/srcTransformer/less_transformer.dart 30:23  LessTransformer.transform.<fn>
dart:async                                                    runZoned
package:less_dart/srcTransformer/less_transformer.dart 29:5   LessTransformer.transform
package:less_dart/transformer.dart 90:32                      FileTransformer.apply.<fn>.<fn>.<fn>
dart:isolate                                                  _RawReceivePortImpl._handleMessage
```

**Comment by AdalbertoLacruz** [comment]
From the source: package:less_dart/srcTransformer/less_transformer.dart  lines 29-46 you could see a random number generator is used. Sounds like 
     `new Random().nextInt(10000)`
is not working. Could you test this function in your system?

**Comment by ewot** [comment]
new Random().nextInt(1000) seems to work fine from my naive occular inspection anyway. :)

If it would give the same value twice during a transformer run, would that cause this problem?

**Comment by ewot** [comment]
tl; dr; Logger.cache isn't cleared, causing id clash (and memory leak)

In my case, running pub from inside Eclipse, the Logger.cache is ever growing. I guess this is because pub keeps running and the same transformer instance is reused.

The Logger.cache is never cleared and eventually there will be an id clash. This will also cause a memory leak.

I don't know much about transformers, but if there is a hook where you can clear the cache after the transformer has finished the current run it will probably fix this problem.

**Comment by AdalbertoLacruz** [comment]
Could you try 0.3.2? (here in github)
The random generator is out, and the cache is reset to avoid memory leaks.

**Comment by ewot** [comment]
The fix works great! Haven't been able to reproduce the issue. 

Thank you!

**Comment by AdalbertoLacruz** [comment]
Updated pub with 0.3.2
